File: ranker_cache.py
import sqlite3
import json
import hashlib
from typing import List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from llm_ranker import LLMRanker


class RankerCache:
    """
    Cache for LLM ranker operations to avoid recomputing expensive token ranking operations.
    
    Stores cached results for:
    - get_token_ranks: text -> ranks
    - get_string_from_token_ranks: ranks -> text
    
    Cache keys are based on ranker signature to ensure different ranker configurations
    don't interfere with each other.
    """
    
    def __init__(self, db_path: str = "compression_results.db"):
        """
        Initialize the ranker cache.
        
        Args:
            db_path: Path to SQLite database for caching
        """
        self.db_path = db_path
        self._setup_cache_tables()
    
    def _setup_cache_tables(self):
        """Initialize the cache tables in the database."""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Create token ranks cache table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS token_ranks_cache (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                text_hash TEXT NOT NULL,
                ranker_signature TEXT NOT NULL,
                text_sample TEXT NOT NULL,
                ranks TEXT NOT NULL,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                UNIQUE(text_hash, ranker_signature)
            )
        """)
        
        # Create string from ranks cache table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS string_from_ranks_cache (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                ranks_hash TEXT NOT NULL,
                ranker_signature TEXT NOT NULL,
                ranks_sample TEXT NOT NULL,
                result_text TEXT NOT NULL,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                UNIQUE(ranks_hash, ranker_signature)
            )
        """)
        
        conn.commit()
        conn.close()
    
    def _get_text_hash(self, text: str) -> str:
        """Generate a hash for the input text."""
        return hashlib.sha256(text.encode('utf-8')).hexdigest()
    
    def _get_ranks_hash(self, ranks: List[int]) -> str:
        """Generate a hash for the input ranks."""
        ranks_str = json.dumps(ranks)
        return hashlib.sha256(ranks_str.encode('utf-8')).hexdigest()
    
    def _get_ranker_signature_str(self, ranker: 'LLMRanker') -> str:
        """Get a string representation of the ranker signature for database storage."""
        signature = ranker.get_cache_signature()
        return json.dumps(signature, sort_keys=True)
    
    def get_cached_token_ranks(self, text: str, ranker: 'LLMRanker') -> Optional[List[int]]:
        """
        Retrieve cached token ranks if available.
        
        Args:
            text: Input text
            ranker: LLM ranker instance
            
        Returns:
            Cached ranks if found, None otherwise
        """
        text_hash = self._get_text_hash(text)
        ranker_signature = self._get_ranker_signature_str(ranker)
        
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute("""
            SELECT ranks FROM token_ranks_cache 
            WHERE text_hash = ? AND ranker_signature = ?
        """, (text_hash, ranker_signature))
        
        result = cursor.fetchone()
        conn.close()
        
        if result:
            return json.loads(result[0])
        return None
    
    def cache_token_ranks(self, text: str, ranker: 'LLMRanker', ranks: List[int]):
        """
        Cache the computed token ranks.
        
        Args:
            text: Input text
            ranker: LLM ranker instance
            ranks: Computed token ranks
        """
        text_hash = self._get_text_hash(text)
        ranker_signature = self._get_ranker_signature_str(ranker)
        ranks_json = json.dumps(ranks)
        # Store first 100 characters of text as sample for debugging
        text_sample = text[:100]
        
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute("""
            INSERT OR REPLACE INTO token_ranks_cache 
            (text_hash, ranker_signature, text_sample, ranks)
            VALUES (?, ?, ?, ?)
        """, (text_hash, ranker_signature, text_sample, ranks_json))
        
        conn.commit()
        conn.close()
    
    def get_cached_string_from_ranks(self, ranks: List[int], ranker: 'LLMRanker') -> Optional[str]:
        """
        Retrieve cached string from ranks if available.
        
        Args:
            ranks: Input ranks
            ranker: LLM ranker instance
            
        Returns:
            Cached string if found, None otherwise
        """
        ranks_hash = self._get_ranks_hash(ranks)
        ranker_signature = self._get_ranker_signature_str(ranker)
        
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute("""
            SELECT result_text FROM string_from_ranks_cache 
            WHERE ranks_hash = ? AND ranker_signature = ?
        """, (ranks_hash, ranker_signature))
        
        result = cursor.fetchone()
        conn.close()
        
        if result:
            return result[0]
        return None
    
    def cache_string_from_ranks(self, ranks: List[int], ranker: 'LLMRanker', result_text: str):
        """
        Cache the computed string from ranks.
        
        Args:
            ranks: Input ranks
            ranker: LLM ranker instance
            result_text: Computed result text
        """
        ranks_hash = self._get_ranks_hash(ranks)
        ranker_signature = self._get_ranker_signature_str(ranker)
        ranks_json = json.dumps(ranks)
        # Store first 100 characters of ranks as sample for debugging
        ranks_sample = ranks_json[:100]
        
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute("""
            INSERT OR REPLACE INTO string_from_ranks_cache 
            (ranks_hash, ranker_signature, ranks_sample, result_text)
            VALUES (?, ?, ?, ?)
        """, (ranks_hash, ranker_signature, ranks_sample, result_text))
        
        conn.commit()
        conn.close()
    
    def get_token_ranks_cached(self, text: str, ranker: 'LLMRanker') -> List[int]:
        """
        Get token ranks with caching.
        
        Args:
            text: Input text
            ranker: LLM ranker instance
            
        Returns:
            Token ranks (from cache or computed)
        """
        # Check cache first
        cached_ranks = self.get_cached_token_ranks(text, ranker)
        if cached_ranks is not None:
            return cached_ranks
        
        # Not in cache, compute using ranker
        ranks = ranker.get_token_ranks(text)
        
        # Cache the result
        self.cache_token_ranks(text, ranker, ranks)
        
        return ranks
    
    def get_string_from_token_ranks_cached(self, ranks: List[int], ranker: 'LLMRanker') -> str:
        """
        Get string from token ranks with caching.
        
        Args:
            ranks: Input ranks
            ranker: LLM ranker instance
            
        Returns:
            Generated text (from cache or computed)
        """
        # Check cache first
        cached_result = self.get_cached_string_from_ranks(ranks, ranker)
        if cached_result is not None:
            return cached_result
        
        # Not in cache, compute using ranker
        result_text = ranker.get_string_from_token_ranks(ranks)
        
        # Cache the result
        self.cache_string_from_ranks(ranks, ranker, result_text)
        
        return result_text
    
    def clear_cache(self):
        """Clear all cached data."""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute("DELETE FROM token_ranks_cache")
        cursor.execute("DELETE FROM string_from_ranks_cache")
        
        conn.commit()
        conn.close()
    
    def get_cache_stats(self) -> dict:
        """
        Get statistics about the cache.
        
        Returns:
            Dictionary with cache statistics
        """
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Get token ranks cache stats
        cursor.execute("SELECT COUNT(*) FROM token_ranks_cache")
        token_ranks_count = cursor.fetchone()[0]
        
        cursor.execute("SELECT COUNT(*) FROM string_from_ranks_cache")
        string_ranks_count = cursor.fetchone()[0]
        
        # Get unique ranker signatures
        cursor.execute("SELECT DISTINCT ranker_signature FROM token_ranks_cache")
        token_signatures = cursor.fetchall()
        
        cursor.execute("SELECT DISTINCT ranker_signature FROM string_from_ranks_cache")
        string_signatures = cursor.fetchall()
        
        conn.close()
        
        return {
            "token_ranks_entries": token_ranks_count,
            "string_from_ranks_entries": string_ranks_count,
            "total_entries": token_ranks_count + string_ranks_count,
            "unique_ranker_signatures_token": len(token_signatures),
            "unique_ranker_signatures_string": len(string_signatures)
        }

File: javascript/src/ranker-cache.js
import { createHash } from 'crypto';

// Try to import sqlite3, fall back to mock if not available
let Database;
try {
    const sqlite3 = await import('sqlite3');
    Database = sqlite3.Database;
} catch (error) {
    // Mock Database class for testing without sqlite3
    Database = class MockDatabase {
        constructor() {}
        run() {}
        get(query, params, callback) { callback(null, null); }
        all(query, params, callback) { callback(null, []); }
        close() {}
    };
}

/**
 * Cache for LLM ranker operations to avoid recomputing expensive operations
 */
export class RankerCache {
    /**
     * Initialize the ranker cache
     * @param {string} dbPath - Path to SQLite database
     */
    constructor(dbPath = 'compression_results.db') {
        this.dbPath = dbPath;
        this._setupCacheTables();
    }

    /**
     * Initialize cache tables in the database
     * @private
     */
    _setupCacheTables() {
        const db = new Database(this.dbPath);
        
        // Create token ranks cache table
        db.run(`
            CREATE TABLE IF NOT EXISTS token_ranks_cache (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                text_hash TEXT NOT NULL,
                ranker_signature TEXT NOT NULL,
                text_sample TEXT NOT NULL,
                ranks TEXT NOT NULL,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                UNIQUE(text_hash, ranker_signature)
            )
        `);
        
        // Create string from ranks cache table
        db.run(`
            CREATE TABLE IF NOT EXISTS string_from_ranks_cache (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                ranks_hash TEXT NOT NULL,
                ranker_signature TEXT NOT NULL,
                ranks_sample TEXT NOT NULL,
                result_text TEXT NOT NULL,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                UNIQUE(ranks_hash, ranker_signature)
            )
        `);
        
        db.close();
    }

    /**
     * Generate a hash for input text
     * @param {string} text - Input text
     * @returns {string} Hash string
     * @private
     */
    _getTextHash(text) {
        return createHash('sha256').update(text, 'utf8').digest('hex');
    }

    /**
     * Generate a hash for input ranks
     * @param {Array<number>} ranks - Input ranks
     * @returns {string} Hash string
     * @private
     */
    _getRanksHash(ranks) {
        const ranksStr = JSON.stringify(ranks);
        return createHash('sha256').update(ranksStr, 'utf8').digest('hex');
    }

    /**
     * Get string representation of ranker signature
     * @param {Object} ranker - Ranker instance
     * @returns {string} Signature string
     * @private
     */
    _getRankerSignatureStr(ranker) {
        const signature = ranker.getCacheSignature();
        return JSON.stringify(signature);
    }

    /**
     * Retrieve cached token ranks if available
     * @param {string} text - Input text
     * @param {Object} ranker - Ranker instance
     * @returns {Promise<Array<number>|null>} Cached ranks or null
     */
    async getCachedTokenRanks(text, ranker) {
        const textHash = this._getTextHash(text);
        const rankerSignature = this._getRankerSignatureStr(ranker);
        
        return new Promise((resolve, reject) => {
            const db = new Database(this.dbPath);
            
            const query = `
                SELECT ranks FROM token_ranks_cache 
                WHERE text_hash = ? AND ranker_signature = ?
            `;
            
            db.get(query, [textHash, rankerSignature], (err, row) => {
                db.close();
                if (err) {
                    reject(err);
                } else if (row) {
                    resolve(JSON.parse(row.ranks));
                } else {
                    resolve(null);
                }
            });
        });
    }

    /**
     * Cache computed token ranks
     * @param {string} text - Input text
     * @param {Object} ranker - Ranker instance
     * @param {Array<number>} ranks - Computed ranks
     * @returns {Promise<void>}
     */
    async cacheTokenRanks(text, ranker, ranks) {
        const textHash = this._getTextHash(text);
        const rankerSignature = this._getRankerSignatureStr(ranker);
        const ranksJson = JSON.stringify(ranks);
        const textSample = text.substring(0, 100);
        
        return new Promise((resolve, reject) => {
            const db = new Database(this.dbPath);
            
            const query = `
                INSERT OR REPLACE INTO token_ranks_cache 
                (text_hash, ranker_signature, text_sample, ranks)
                VALUES (?, ?, ?, ?)
            `;
            
            db.run(query, [textHash, rankerSignature, textSample, ranksJson], function(err) {
                db.close();
                if (err) {
                    reject(err);
                } else {
                    resolve();
                }
            });
        });
    }

    /**
     * Retrieve cached string from ranks if available
     * @param {Array<number>} ranks - Input ranks
     * @param {Object} ranker - Ranker instance
     * @returns {Promise<string|null>} Cached string or null
     */
    async getCachedStringFromRanks(ranks, ranker) {
        const ranksHash = this._getRanksHash(ranks);
        const rankerSignature = this._getRankerSignatureStr(ranker);
        
        return new Promise((resolve, reject) => {
            const db = new Database(this.dbPath);
            
            const query = `
                SELECT result_text FROM string_from_ranks_cache 
                WHERE ranks_hash = ? AND ranker_signature = ?
            `;
            
            db.get(query, [ranksHash, rankerSignature], (err, row) => {
                db.close();
                if (err) {
                    reject(err);
                } else if (row) {
                    resolve(row.result_text);
                } else {
                    resolve(null);
                }
            });
        });
    }

    /**
     * Cache computed string from ranks
     * @param {Array<number>} ranks - Input ranks
     * @param {Object} ranker - Ranker instance
     * @param {string} resultText - Computed result text
     * @returns {Promise<void>}
     */
    async cacheStringFromRanks(ranks, ranker, resultText) {
        const ranksHash = this._getRanksHash(ranks);
        const rankerSignature = this._getRankerSignatureStr(ranker);
        const ranksJson = JSON.stringify(ranks);
        const ranksSample = ranksJson.substring(0, 100);
        
        return new Promise((resolve, reject) => {
            const db = new Database(this.dbPath);
            
            const query = `
                INSERT OR REPLACE INTO string_from_ranks_cache 
                (ranks_hash, ranker_signature, ranks_sample, result_text)
                VALUES (?, ?, ?, ?)
            `;
            
            db.run(query, [ranksHash, rankerSignature, ranksSample, resultText], function(err) {
                db.close();
                if (err) {
                    reject(err);
                } else {
                    resolve();
                }
            });
        });
    }

    /**
     * Get token ranks with caching
     * @param {string} text - Input text
     * @param {Object} ranker - Ranker instance
     * @returns {Promise<Array<number>>} Token ranks
     */
    async getTokenRanksCached(text, ranker) {
        // Check cache first
        const cachedRanks = await this.getCachedTokenRanks(text, ranker);
        if (cachedRanks !== null) {
            return cachedRanks;
        }
        
        // Not in cache, compute using ranker
        const ranks = await ranker.getTokenRanks(text);
        
        // Cache the result
        await this.cacheTokenRanks(text, ranker, ranks);
        
        return ranks;
    }

    /**
     * Get string from token ranks with caching
     * @param {Array<number>} ranks - Input ranks
     * @param {Object} ranker - Ranker instance
     * @returns {Promise<string>} Generated text
     */
    async getStringFromTokenRanksCached(ranks, ranker) {
        // Check cache first
        const cachedResult = await this.getCachedStringFromRanks(ranks, ranker);
        if (cachedResult !== null) {
            return cachedResult;
        }
        
        // Not in cache, compute using ranker
        const resultText = await ranker.getStringFromTokenRanks(ranks);
        
        // Cache the result
        await this.cacheStringFromRanks(ranks, ranker, resultText);
        
        return resultText;
    }

    /**
     * Clear all cached data
     * @returns {Promise<void>}
     */
    async clearCache() {
        return new Promise((resolve, reject) => {
            const db = new Database(this.dbPath);
            
            db.serialize(() => {
                db.run('DELETE FROM token_ranks_cache');
                db.run('DELETE FROM string_from_ranks_cache', function(err) {
                    db.close();
                    if (err) {
                        reject(err);
                    } else {
                        resolve();
                    }
                });
            });
        });
    }

    /**
     * Get cache statistics
     * @returns {Promise<Object>} Cache statistics
     */
    async getCacheStats() {
        return new Promise((resolve, reject) => {
            const db = new Database(this.dbPath);
            
            const stats = {};
            
            db.serialize(() => {
                db.get('SELECT COUNT(*) as count FROM token_ranks_cache', (err, row) => {
                    if (err) {
                        db.close();
                        reject(err);
                        return;
                    }
                    stats.tokenRanksEntries = row.count;
                });
                
                db.get('SELECT COUNT(*) as count FROM string_from_ranks_cache', (err, row) => {
                    if (err) {
                        db.close();
                        reject(err);
                        return;
                    }
                    stats.stringFromRanksEntries = row.count;
                });
                
                db.all('SELECT DISTINCT ranker_signature FROM token_ranks_cache', (err, rows) => {
                    if (err) {
                        db.close();
                        reject(err);
                        return;
                    }
                    stats.uniqueRankerSignaturesToken = rows.length;
                });
                
                db.all('SELECT DISTINCT ranker_signature FROM string_from_ranks_cache', (err, rows) => {
                    db.close();
                    if (err) {
                        reject(err);
                        return;
                    }
                    stats.uniqueRankerSignaturesString = rows.length;
                    stats.totalEntries = stats.tokenRanksEntries + stats.stringFromRanksEntries;
                    resolve(stats);
                });
            });
        });
    }
}